package/samplers/auto_sampler/sampler.py

```diff
@@ -0,0 +1,243 @@
+            be a sequence of :obj:`float` s. A value strictly larger than 0 means that a
+            constraints is violated. A value equal to or smaller than 0 is considered feasible.
+            If ``constraints_func`` returns more than one value for a trial, that trial is
+            considered feasible if and only if all values are equal to 0 or smaller.
+
+            The ``constraints_func`` will be evaluated after each successful trial.
+            The function won't be called when trials fail or they are pruned, but this behavior is
+            subject to change in the future releases.
+    """
+
+    def __init__(
+        self,
+        seed: int | None = None,
+        constraints_func: Callable[[FrozenTrial], Sequence[float]] | None = None,
+    ) -> None:
+        self._rng = LazyRandomState(seed)
+        seed_for_random_sampler = self._rng.rng.randint(MAXINT32)
+        self._sampler: BaseSampler = RandomSampler(seed=seed_for_random_sampler)
+        self._constraints_func = constraints_func
+
+    def reseed_rng(self) -> None:
+        self._rng.rng.seed()
+        self._sampler.reseed_rng()
+
+    def _include_conditional_param(self, study: Study) -> bool:
+        trials = study.get_trials(deepcopy=False, states=(TrialState.COMPLETE, TrialState.PRUNED))
+        if len(trials) == 0:
+            return False
+
+        param_key = set(trials[0].params)
+        for t in trials:
+            if param_key != set(t.params):
+                return True
+
+        return False
+
+    def _determine_multi_objective_sampler(
+        self, study: Study, trial: FrozenTrial, search_space: dict[str, BaseDistribution]
+    ) -> None:
+        if isinstance(self._sampler, (NSGAIISampler, NSGAIIISampler)):
+            return
+
+        seed = self._rng.rng.randint(MAXINT32)
+        if not isinstance(self._sampler, TPESampler):
+            self._sampler = TPESampler(
+                seed=seed,
+                multivariate=True,
+                warn_independent_sampling=False,
+                constraints_func=self._constraints_func,
+                constant_liar=True,
+            )
+            return
+
+        complete_trials = study.get_trials(deepcopy=False, states=(TrialState.COMPLETE,))
+        complete_trials.sort(key=lambda trial: trial.datetime_complete)
+        if len(complete_trials) >= self._N_COMPLETE_TRIALS_FOR_NSGA:
+            nsga_sampler_cls = (
+                NSGAIISampler
+                if len(study.directions) < THRESHOLD_OF_MANY_OBJECTIVES
+                else NSGAIIISampler
+            )
+            # Use NSGA-II/III if len(complete_trials) <= _N_COMPLETE_TRIALS_FOR_NSGA.
+            self._sampler = nsga_sampler_cls(constraints_func=self._constraints_func, seed=seed)
+
+    def _determine_single_objective_sampler(
+        self, study: Study, trial: FrozenTrial, search_space: dict[str, BaseDistribution]
+    ) -> None:
+        if isinstance(self._sampler, TPESampler):
+            return
```

Comment:
Without `threading.Lock`, these logic looks unsafe since `self._sampler` may be replaced after L135 by another threads. What do you think?

Reply:
By making `self._sampler` threadlocal, this issue has been resolved.